subsys/instrumentation/common/instr_common.c

```diff
@@ -0,0 +1,567 @@
+	 * when stopper function exits. This two step mechanism allow disabling instrumentation at
+	 * runtime in critical sections inside the instrumentation code by calling instr_disable()/
+	 * instr_enable() at runtime, hence avoiding infinite loop and having to exclude a function
+	 * at compile time. In other words, it allows a better granularity for enabling/disabling
+	 * the instrumentation.
+	 */
+	instr_enable();
+
+	return 0;
+}
+
+__no_instrumentation__
+bool instr_initialized(void)
+{
+	return _instr_initialized;
+}
+
+/*
+ * Instrumentation can only be used when RAM is correctly initialized in early
+ * boot stages and so variables in memory -- for example, _instr_initialized --
+ * are correctly initialized. To ensure such a condition, a given variable,
+ * 'magic', has its value in memory checked against a constant (a magic number)
+ * that is kept in code (flash). Once the value matches the constant it means
+ * the RAM is correctly initialized and so instrumentation can be initialized
+ * properly and enabled for use.
+ */
+__no_instrumentation__
+bool instr_fundamentals_initialized(void)
+{
+	if (magic == 0xABBA) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+__no_instrumentation__
+int instr_enable(void)
+{
+	_instr_enabled = true;
+
+	return 0;
+}
+
+__no_instrumentation__
+int instr_disable(void)
+{
+	_instr_enabled = false;
+
+	return 0;
+}
+
+__no_instrumentation__
+bool instr_enabled(void)
+{
+	return _instr_enabled;
+}
+
+__no_instrumentation__
+int instr_turn_on(void)
+{
+	_instr_on = true;
+
+	return 0;
+}
+
+__no_instrumentation__
+int instr_turn_off(void)
+{
+	_instr_on = false;
+
+	return 0;
+}
+
+__no_instrumentation__
+bool instr_turned_on(void)
+{
+	if (!_instr_enabled) {
+		/* If instrumentation is disabled always return off state */
+		return false;
+	} else {
+		return _instr_on;
+	}
+}
+
+__no_instrumentation__
+bool instr_trace_enabled(void)
+{
+	return !_instr_tracing_disabled;
+}
+
+__no_instrumentation__
+bool instr_profile_enabled(void)
+{
+	return !_instr_profiling_disabled;
+}
+
+__no_instrumentation__
+void instr_set_trigger_func(void *callee)
+{
+	/* Update trigger_callee before updating retained mem */
+	trigger_callee = callee;
+
+	retention_write(retention0, 0, (const uint8_t *)&trigger_callee,
+			sizeof(trigger_callee));
+}
+
+__no_instrumentation__
+void instr_set_stop_func(void *callee)
+{
+	/* Update stopper_callee before updatting retained mem */
+	stopper_callee = callee;
+
+	retention_write(retention0, sizeof(trigger_callee), (const uint8_t *)&stopper_callee,
+			sizeof(stopper_callee));
+}
+
+__no_instrumentation__
+void *instr_get_trigger_func(void)
+{
+	return trigger_callee;
+}
+
+__no_instrumentation__
+void *instr_get_stop_func(void)
+{
+	return stopper_callee;
+}
+
+__no_instrumentation__
+void instr_dump_buffer_uart(void)
+{
+#if defined(CONFIG_INSTRUMENTATION_MODE_CALLGRAPH)
+	static const struct device *const uart_dev =
+	DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
```

Comment:
Maybe in a future enhancement this could also use an alias, so that a user can have 2 UARTs which allows the console to still be used and have instrumentation data dumped on a dedicated UART